src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer_GetDiagnostics.cs

```diff
@@ -428,31 +453,25 @@ protected override Task AppendProjectStateDiagnosticsAsync(Project project, Docu
 
             private async Task<CompilationWithAnalyzers> GetCompilationWithAnalyzersAsync(Project project, bool concurrentAnalysis, bool reportSuppressedDiagnostics, CancellationToken cancellationToken)
             {
-                CompilationWithAnalyzers compilationWithAnalyzers;
-                lock (_compilationWithAnalyzersMap)
-                {
-                    if (_compilationWithAnalyzersMap.TryGetValue(project.Id, out compilationWithAnalyzers))
-                    {
-                        return compilationWithAnalyzers;
-                    }
-                }
-
-                if (project.SupportsCompilation)
+                if (_compilationWithAnalyzers == null && project.SupportsCompilation)
                 {
                     var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
-                    compilationWithAnalyzers = Owner.GetCompilationWithAnalyzers(project, compilation, concurrentAnalysis, reportSuppressedDiagnostics);
-                }
-                else
-                {
-                    compilationWithAnalyzers = null;
+                    var analyzers = GetAnalyzers(project);
+
+                    _compilationWithAnalyzers = Owner.GetCompilationWithAnalyzers(project, analyzers, compilation, concurrentAnalysis, reportSuppressedDiagnostics);
```

Comment:
This can be invoked from multiple threads as latest diagnostic getter is invoked in fix all code path which invokes AppendDiagnostics in parallel for all documents. You need lock statements.